projects/feature-activation/0004-phased-testing.md

```diff
@@ -0,0 +1,209 @@
+3. After 2 weeks, it gets to `MUST_SIGNAL` on block `N + 40320`, for its `lock_in_on_timeout` configuration.
+4. After 2 weeks, it gets to `LOCKED_IN` on block  `N + 2 * 40320`.
+5. After 2 weeks, it gets to `ACTIVE` on block `N + 3 * 40320`.
+
+Therefore, `NOP_FEATURE_2`'s process takes 6 weeks.
+
+#### `NOP_FEATURE_3`
+
+1. It begins as `DEFINED`, up to block `N - 1`.
+2. It transitions to `STARTED` on block `N`, for reaching its `start_height`.
+3. After 2 weeks, it continues on `STARTED` on block `N + 40320`, as its `timeout_height` has not been reached yet.
+4. After 2 weeks, it gets to `FAILED` on block  `N + 2 * 40320`, as its `timeout_height` is reached and `lock_in_on_timeout` is `false`.
+
+Therefore, `NOP_FEATURE_3`'s process takes 4 weeks.
+
+### Validation via the Hathor Explorer
+
+By manually inspecting blocks via the Hathor Explorer interfaces described in the [Feature Activation Explorer UIs RFC](https://github.com/HathorNetwork/rfcs/blob/master/projects/feature-activation/0003-explorer-uis.md), we'll be able to validate the expected state for each block-feature pair according to the lists described above.
+
+### Validation via logging
+
+Complementing the validation via Explorer, we'll also have validation via full node logs, which will be useful to perform more complex queries to understand state transitions, specially if debugging is necessary. By using structured logging of features states for each block, we'll be able to leverage AWS CloudWatch queries to perform analysis and even export metrics.
+
+After a new block is received and processed by the full node, we'll log its height, together with its state for each NOP feature. At the end of the `HathorManager.tx_fully_validated()` method, we'll call a new `HathorManager.log_feature_states()` method, defined as follows:
+
+```python
+def log_feature_states(self, vertex: BaseTransaction) -> None:
+    if not isinstance(vertex, Block):
+        return
+
+    feature_descriptions = self.feature_service.get_bits_description(block=vertex)
+    state_by_feature = {
+        feature.value: description.state.value
+        for feature, description in feature_descriptions.items()
+    }
+
+    self.log.info(
+        'New block accepted with feature activation states',
+        block_height=vertex.get_height(),
+        features_states=state_by_feature
+    )
+```
+
+This logging will be removed after Phase 1 is complete.
+
+### Phase 1b
+
+After the initial part of Phase 1 is complete, Phase 1b will start. On this phase, we'll force a reorg on testnet to undo the activation of `NOP_FEATURE_1`. We'll have to disable support of said feature in our miners and use a higher hash rate to generate more blocks than the original best chain, using some block before the 75% threshold is reached as the common block. After the procedure is complete, we must observe that `NOP_FEATURE_1` transitions back to `FAILED` instead of `ACTIVE`.
```

Comment:
Should we run a separate private test network to test this part of the implementation? I mean, it's easy to cause a reorg on the testnet but maybe we are better off running this test in a private network.

Reply:
We could, but I would also like to see this behavior happening in an environment as similar as possible to our mainnet, since we'll not test this on mainnet, of course.

If the testnet is too sensitive for this test, maybe we should have a new running network for development, like a dev-testnet, where we can freely do tests. What do you think? Maybe what I'm suggesting is just the same as a one-off private network...

Reply:
I guess a dev network would be the same as a private network since we will control all peers, miners, and wallets connected to it.

Anyway, it's ok to use the public testnet for this. As we mine most of the blocks, we should be able to activate the feature as we wish. We should let the community know that we are running this test since they are supposed to upgrade their full nodes too.